crates/miden-objects/src/block/account_tree.rs

```diff
@@ -119,17 +142,79 @@ impl AccountTree {
         AccountWitness::from_smt_proof(account_id, proof)
     }
 
+    /// Returns an opening of the leaf associated with the `account_id` at a specific block.
+    ///
+    /// # Arguments
+    ///
+    /// * `account_id` - The account ID to query
+    /// * `block_number` - The block number to query
+    ///
+    /// # Returns
+    ///
+    /// Returns `None` if the requested block is not available in history.
+    pub fn open_at(
+        &self,
+        account_id: AccountId,
+        block_number: BlockNumber,
+    ) -> Option<AccountWitness> {
+        let historical_view = self.smt.historical_view(block_number.as_u64())?;
+        let key = Self::id_to_smt_key(account_id);
+        let proof = historical_view.open(&key);
+
+        Some(AccountWitness::from_smt_proof(account_id, proof))
+    }
+
     /// Returns the current state commitment of the given account ID.
     pub fn get(&self, account_id: AccountId) -> Word {
         let key = Self::id_to_smt_key(account_id);
         self.smt.get_value(&key)
     }
 
+    /// Returns the state commitment of the given account ID at a specific block.
+    ///
+    /// # Arguments
+    ///
+    /// * `account_id` - The account ID to query
+    /// * `block_number` - The block number to query
+    ///
+    /// # Returns
+    ///
+    /// Returns `None` if the requested block is not available in history.
+    pub fn get_at(&self, account_id: AccountId, block_number: BlockNumber) -> Option<Word> {
+        let key = Self::id_to_smt_key(account_id);
+        let historical_view = self.smt.historical_view(block_number.as_u64())?;
+        Some(historical_view.get_value(&key))
+    }
+
     /// Returns the root of the tree.
     pub fn root(&self) -> Word {
         self.smt.root()
     }
 
+    /// Returns the root of the tree at a specific block.
+    ///
+    /// # Arguments
+    ///
+    /// * `block_number` - The block number to query
+    ///
+    /// # Returns
+    ///
+    /// Returns `None` if the requested block is not available in history.
+    pub fn root_at(&self, block_number: BlockNumber) -> Option<Word> {
+        let historical_view = self.smt.historical_view(block_number.as_u64())?;
+        Some(historical_view.root())
+    }
+
+    /// Returns the number of historical states currently maintained.
+    pub fn history_len(&self) -> usize {
+        self.smt.history_len()
+    }
+
+    /// Returns the oldest offset still available in history.
+    pub fn oldest_offset(&self) -> usize {
+        self.smt.oldest()
+    }
```

Comment:
So if we have a tree at block 20 and we retain 5 historical states, then `history_len` would return `5` and `oldest_offset` would return `15`? If so, I would suggest renaming `oldest_offset` to `oldest_block_num` to match `current_block_num`/`latest_block_num` and return `BlockNumber`.